src/conda_package_handling/archive_utils_c.c

```diff
@@ -0,0 +1,186 @@
+        *err_str = archive_error_string(a);
+        archive_write_close(a);
+        archive_write_free(a);
+        return NULL;
+    }
+    if (archive_write_set_options(a, opts) < ARCHIVE_OK) {
+        *err_str = archive_error_string(a);
+        archive_write_close(a);
+        archive_write_free(a);
+        return NULL;
+    }
+	if (archive_write_open_filename(a, outname) < ARCHIVE_OK) {
+        *err_str = archive_error_string(a);
+        archive_write_close(a);
+        archive_write_free(a);
+        return NULL;
+    }
+	return a;
+}
+
+void close_archive(struct archive *a) {
+	archive_write_close(a);
+	archive_write_free(a);
+}
+
+struct archive_entry * prepare_entry(void) {
+	struct archive_entry *entry;
+    entry = archive_entry_new();
+    return entry;
+}
+
+void close_entry(struct archive_entry *entry) {
+    archive_entry_free(entry);
+}
+
+static int add_file(
+    struct archive *a, struct archive_entry *entry, const char *filename, const char **err_str)
+{
+    struct archive *disk;
+	char buff[8192];
+	int len;
+	int fd;
+    int flags;
+    flags = 0;
+
+    disk = archive_read_disk_new();
+    if (disk == NULL) {
+        return 1;
+    }
+    if (archive_read_disk_set_behavior(disk, flags) < ARCHIVE_OK) {
+        *err_str = archive_error_string(disk);
+        return 1;
+    }
+    if (archive_read_disk_open(disk, filename) < ARCHIVE_OK) {
+        *err_str = archive_error_string(disk);
+        return 1;
+    }
+    if (archive_read_next_header2(disk, entry) < ARCHIVE_OK) {
+        *err_str = archive_error_string(disk);
+        return 1;
+    }
+    if (archive_read_disk_descend(disk) < ARCHIVE_OK) {
+        *err_str = archive_error_string(disk);
+        return 1;
+    }
+    if (archive_write_header(a, entry) < ARCHIVE_OK) {
+        *err_str = archive_error_string(a);
+        return 1;
+    }
+    fd = open(filename, O_RDONLY);
+    len = read(fd, buff, sizeof(buff));
+    while ( len > 0 ) {
+        archive_write_data(a, buff, len);
+        len = read(fd, buff, sizeof(buff));
+    }
+    close(fd);
+    if (archive_write_finish_entry(a) < ARCHIVE_OK) {
+        *err_str = archive_error_string(a);
+        return 1;
+    }
+    archive_read_close(disk);
+    archive_read_free(disk);
+    archive_entry_clear(entry);
+    return 0;
+}
+
+
+static int copy_data(struct archive *ar, struct archive *aw)
+{
+    int r;
+    const void *buff;
+    size_t size;
+    la_int64_t offset;
+    for (;;) {
+        r = archive_read_data_block(ar, &buff, &size, &offset);
+        if (r == ARCHIVE_EOF)
+            return (ARCHIVE_OK);
+        if (r < ARCHIVE_OK)
+            return (r);
+        r = archive_write_data_block(aw, buff, size, offset);
+        if (r < ARCHIVE_OK) {
+            return (r);
+    }
+  }
+}
+
+static int extract_file_c(const char *filename, const char **err_str) {
+    struct archive *a;
+    struct archive *ext;
+    struct archive_entry *entry;
+    int flags;
+    int r;
+
```

Comment:
Same here about err_str. check required that err_str is not NULL.  That helps to avoid warnings